lxd/device/gpu_physical.go

```diff
@@ -342,14 +537,69 @@ func (d *gpuPhysical) pciDeviceDriverOverrideIOMMU(pciDev pcidev.Device, driverO
 	return nil
 }
 
+// stopCDIDevices reads the configDevices and remove potential unix device and unmounts disk mounts.
+func (d *gpuPhysical) stopCDIDevices(configDevices cdi.ConfigDevices, runConf *deviceConfig.RunConfig) error {
+	// Remove ALL the underlying unix-char dev entries created when the CDI device started.
+	err := unixDeviceRemove(d.inst.DevicesPath(), "cdi.unix", d.name, "", runConf)
+	if err != nil {
+		return err
+	}
+
+	for _, conf := range configDevices.BindMounts {
+		relativeDestPath := strings.TrimPrefix(conf["path"], "/")
+		devPath := filepath.Join(d.inst.DevicesPath(), filesystem.PathNameEncode(deviceJoinPath("cdi.disk", d.name, relativeDestPath)))
+		runConf.PostHooks = append(runConf.PostHooks, func() error {
+			// Clean any existing device mount entry. Should occur first before custom volume unmounts.
+			err := DiskMountClear(devPath)
+			if err != nil {
+				return err
+			}
+
+			return nil
+		})
+
+		// The disk device doesn't exist do nothing.
+		if !shared.PathExists(devPath) {
+			return nil
+		}
+
+		// Request an unmount of the device inside the instance.
+		runConf.Mounts = append(runConf.Mounts, deviceConfig.MountEntryItem{
+			TargetPath: relativeDestPath,
+		})
+	}
+
+	return nil
+}
+
 // Stop is run when the device is removed from the instance.
 func (d *gpuPhysical) Stop() (*deviceConfig.RunConfig, error) {
 	runConf := deviceConfig.RunConfig{
 		PostHooks: []func() error{d.postStop},
 	}
 
 	if d.inst.Type() == instancetype.Container {
-		err := unixDeviceRemove(d.inst.DevicesPath(), "unix", d.name, "", &runConf)
+		cdiID, err := cdi.ToCDI(d.config["id"])
+		// In case of an 'id' not being CDI-compliant (e.g, a legacy DRM card id), we remove unix devices only as usual.
+		if err == nil && cdiID.Empty() {
+			err := unixDeviceRemove(d.inst.DevicesPath(), "unix", d.name, "", &runConf)
+			if err != nil {
+				return nil, err
+			}
+
+			return &runConf, nil
+		}
+
+		if err != nil {
+			return nil, err
+		}
+
+		configDevices, _, err := cdi.GenerateFromCDI(d.inst, cdiID, d.logger)
```

Comment:
Can you keep the structure of this function consistent with how you did it in startContainer and instead of handling the previous method first, do that by default and handle the CDI method by exception ,e.g

```go
if !cdiID.Empty() { 
   configDevices, _, err := cdi.GenerateFromCDI(d.inst, cdiID, d.logger)
   ...
}
```

Reply:
fixed